src/libAtomVM/context.c

```diff
@@ -112,53 +115,141 @@ Context *context_new(GlobalContext *glb)
+    } else {
+        ctx->exit_reason = memory_copy_term_tree(&ctx->heap_ptr, signal->term, &ctx->mso_list);
+    }
+    context_update_flags(ctx, ~NoFlags, Killed);
+}
 
-    struct ListHead *item;
-    LIST_FOR_EACH (item, &ctx->mailbox) {
-        num_messages++;
+void context_process_process_info_request_signal(Context *ctx, struct BuiltInAtomRequestSignal *signal)
+{
+    Context *target = globalcontext_get_process_lock(ctx->global, signal->sender_pid);
+    if (target) {
+        term ret;
+        if (context_get_process_info(ctx, &ret, signal->atom)) {
+            mailbox_send_term_signal(target, TrapAnswerSignal, ret);
+        } else {
+            mailbox_send_built_in_atom_signal(target, TrapExceptionSignal, ret);
+        }
+        globalcontext_get_process_unlock(ctx->global, target);
+    } // else: sender died
+}
+
+bool context_process_signal_trap_answer(Context *ctx, struct TermSignal *signal)
+{
+    context_update_flags(ctx, ~Trap, NoFlags);
+    if (UNLIKELY(memory_ensure_free(ctx, signal->msg_memory_size))) {
+        return false;
     }
+    ctx->x[0] = memory_copy_term_tree(&ctx->heap_ptr, signal->term, &ctx->mso_list);
+    return true;
+}
 
-    return num_messages;
+void context_update_flags(Context *ctx, int mask, int value)
+{
+    SPIN_LOCK(ctx->flags_lock);
+    ctx->flags = (ctx->flags & mask) | value;
+    SPIN_UNLOCK(ctx->flags_lock);
 }
 
-size_t context_size(Context *ctx)
+size_t context_message_queue_len(Context *ctx)
 {
-    size_t messages_size = 0;
+    return mailbox_len(&ctx->mailbox);
+}
 
-    struct ListHead *item;
-    LIST_FOR_EACH (item, &ctx->mailbox) {
-        Message *msg = GET_LIST_ENTRY(item, Message, mailbox_list_head);
-        messages_size += sizeof(Message) + msg->msg_memory_size;
-    }
+size_t context_size(Context *ctx)
+{
+    size_t messages_size = mailbox_size(&ctx->mailbox);
 
     // TODO include ctx->platform_data
     return sizeof(Context)
         + messages_size
         + context_memory_size(ctx) * BYTES_PER_TERM;
 }
 
+bool context_gc(Context *ctx)
+{
+    size_t memory_size = context_memory_size(ctx);
+    if (UNLIKELY(memory_gc(ctx, memory_size + MIN_FREE_SPACE_SIZE) != MEMORY_GC_OK)) {
+        return false;
+    }
+
+    size_t free_space = context_avail_free_memory(ctx);
+    size_t minimum_free_space = 2 * MIN_FREE_SPACE_SIZE;
+    if (free_space > minimum_free_space) {
+        memory_size = context_memory_size(ctx);
+        if (UNLIKELY(memory_gc(ctx, (memory_size - free_space) + minimum_free_space) != MEMORY_GC_OK)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool context_get_process_info(Context *ctx, term *out, term atom_key)
+{
+    if (UNLIKELY(memory_ensure_free(ctx, 3) != MEMORY_GC_OK)) {
+        *out = OUT_OF_MEMORY_ATOM;
+        return false;
+    }
+
+    term ret = term_alloc_tuple(2, ctx);
+    // heap_size size in words of the heap of the process
+    if (atom_key == HEAP_SIZE_ATOM) {
```

Comment:
HEAP_SIZE_ATOM and all the other are defined as constants,
hence we can do the following:
```
switch (atom_key) {
    case HEAP_SIZE_ATOM:
[...]
}
```